pkg/api/server/v1alpha2/lister/aggregator.go

```diff
@@ -0,0 +1,208 @@
+	var err error
+	db = db.WithContext(ctx)
+
+	for _, builder := range a.queryBuilders {
+		db, err = builder.build(db)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+	}
+	return db, err
+}
+
+// Aggregate function runs the aggregation tasks and returns Summary
+func (a *Aggregator) Aggregate(ctx context.Context, db *gorm.DB) (*pb.Summary, error) {
+	var err error
+	summary := make([]map[string]interface{}, 0)
+	db = db.Model(&tdb.Record{})
+	db, err = a.buildQuery(ctx, db)
+	if err != nil {
+		return nil, err
+	}
+
+	db = a.applyAggregateFunc(ctx, db)
+	db.Scan(&summary)
+
+	sm, err := ToSummary(summary)
+	if err != nil {
+		return nil, err
+	}
+	return sm, nil
+}
+
+// ToSummary converts the array of summary map to Summary proto
+func ToSummary(summary []map[string]interface{}) (*pb.Summary, error) {
+	var data []*structpb.Struct
+	for _, s := range summary {
+		m := make(map[string]*structpb.Value)
+		for sk, sv := range s {
+			pbValue, err := structpb.NewValue(sv)
+			if err != nil {
+				return nil, err
+			}
+			m[sk] = pbValue
+		}
+		data = append(data, &structpb.Struct{Fields: m})
+	}
+
+	return &pb.Summary{
+		Summary: data,
+	}, nil
+}
+
+var validGroups = map[string]struct{}{
+	"year":   {},
+	"month":  {},
+	"week":   {},
+	"day":    {},
+	"hour":   {},
+	"minute": {},
+}
+
+func newAggregator(env *cel.Env, aggregateObjectRequest summaryRequest, clauses ...equalityClause) (*Aggregator, error) {
+	filter := &filter{
+		env:             env,
+		expr:            strings.TrimSpace(aggregateObjectRequest.GetFilter()),
+		equalityClauses: clauses,
+	}
+
+	summary := strings.Split(strings.TrimSpace(aggregateObjectRequest.GetSummary()), ",")
+	if len(summary) == 0 {
+		// make total default
+		summary = append(summary, "total")
+	}
+
+	aggregators, err := getAggregateFunc(summary)
+	if err != nil {
+		return nil, err
+	}
+
+	group := strings.TrimSpace(aggregateObjectRequest.GetGroupBy())
+	if group != "" {
+		_, ok := validGroups[group]
+		if !ok {
+			return nil, status.Errorf(codes.InvalidArgument, "group_by does not recognize %s", group)
+		}
+		aggregators = append(aggregators, groupBy(group, groupQuery))
+	}
+
+	return &Aggregator{
+		aggregators: aggregators,
+		queryBuilders: []queryBuilder{
+			filter,
+		},
+	}, nil
+}
+
+func getDuration(fn, query, value string) aggregateFunc {
+	return func(db *gorm.DB) *gorm.DB {
+		return db.Select(db.Statement.Selects, fn+"("+query+")::INTERVAL as "+value)
+	}
+}
+
+func groupBy(group, query string) aggregateFunc {
+	return func(db *gorm.DB) *gorm.DB {
+		groupSelect := "EXTRACT(EPOCH FROM DATE_TRUNC('" + group + "', " + query + ")) AS group_value"
+		return db.Select(db.Statement.Selects, groupSelect).Group("group_value")
+	}
+}
+
+func getStatus(query, reason string) aggregateFunc {
+	return func(db *gorm.DB) *gorm.DB {
+		statusSelect := ""
+		if reason == "Unknown" {
```

Comment:
Is there an established definition of `Unknown`? As per tekton pipelines API, there is another status `TaskrunTimeout` and `PipelinerunTimeout`. Moreover PipelineRun API have and extra status `Completed`, when the pipelinerun completes skipped tasks.

Reply:
We are only choosing certain statues. So we can omit other fields.